### DISCUSSION TITLE
Every page results in 404 but only from cold start. Anyone else?

### DISCUSSION THREAD

**Discussion by jesper-bylund** [vercel/next.js#70582] [App Router]
I'm having real issues debugging this since it only seems to happen on cold starts. If prod or local has not had visits for "a few hours" ANY path resolves as 404.
However once refreshed all routes will work perfectly again. I have to wait some random amount of time to see the error again. This happens even with rebooting the server locally.

I've added logging to the middleware and everything is routing fine. Just always resolving the first call as 404.

Has anyone seen this? Anyone know how to debug?

**Comment by romhenri** [comment]
In Next.js, the 404 on cold start could be due to slow route initialization after inactivity. Ensure proper use of `getStaticProps` or `getServerSideProps` and check if your server fully initializes before handling requests. If deploying on Vercel or similar, review any related cold start configurations.

**Reply by jesper-bylund** [reply]
That's super interesting. Is this still true in Nextjs14 with SSR?

**Reply by romhenri** [reply]
Yes.

**Comment by jesper-bylund** [comment]
Apparently not a Nextjs error, just obfuscated by Nextjs routing.
What's happening is the locale of the next-international pacakge is not getting set, resulting in a 404. Why this happens seem to be related to some sort of nextjs cold start though.

`app/[locale]/page.tsx` -> 404, but on refresh 200

**Comment by jesper-bylund** [comment]
For anyone else struggling with this. It turns out that next-international doesn't handle a default locale. It's supposed to. But in this combo of versions:
```
"next": "14.2.13",
"next-international": "1.2.4",
```
you have to manually set a default local in your root layout component.

**Comment by jesper-bylund** [comment]
Turns out my investigation in next-international was completely wrong. I'm still getting 404's from cold start.

What seems to be happening is that Nextjs is unable to handle routing from cold start. Why this is is unclear. My middleware does contain a supabase call. So maybe there's a cold starting lambda to make this call that is timing out?

**Comment by icyJoseph** [comment]
Where is this app deployed to? Is it using App Router or Pages Router? If its a timeout at play, does the 404 take "longer" to resolve?

**Comment by jesper-bylund** [comment]
@icyJoseph  This is deployed to Vercel. It's using App router. And yes the 404 takes a few seconds to resolve.

I found a work around that is highly annoying, following this guide from supabase will result in a 404 from cold start: https://supabase.com/docs/guides/auth/server-side/nextjs

But if the landing page resolves without supabase, everything works fine. 

This should mean that cold starting Nextjs on Vercel takes so long that verifying auth from supabase causes a timeout. Making the router resolve as a 404. I didn't find a way to debug this, and for obvious reasons testing is difficult.

**Reply by icyJoseph** [reply]
Aha, is it the middleware, or just any page that uses the client?

**Reply by jesper-bylund** [reply]
It's the middleware. This should work, but on cold start it fails. And there's no real way to debug since this happens in Nextjs router.

**Reply by icyJoseph** [reply]
And you are 100% sure, it has nothing to do with loading the SDK, but rather, finding the user the first time? Could that be a problem with the location of the DB? or perhaps it itself goes to sleep after a time of inactivity?

**Reply by matejm** [reply]
Anyone got more details on this? We are facing the same issue (also using Supabase and next-international). Not initializing Supabase on landing page doesn't solve the issue, as users can navigate directly to URLs where authentication is required and still get a 404.

**Reply by jesper-bylund** [reply]
I solved this eventually, but I honestly don't know how. I just moved things around in the middleware until suddenly it started working. What I think may have happened is that the middleware times out while waiting for supabase to respond, and therefore sends a 404. But again, I sadly don't know.

**Reply by icyJoseph** [reply]
I learned that undici fetch, the one Node.js implemented, has a rather low timeout of 300 seconds, but I guess that's not the issue here

**Reply by brecht-p7s** [reply]
I had the same issue, it has nothing to do with cold starts! Problem is when the Access Token JWT from Supabase is expired, it calls the `setAll` method passed to the Supabase server client. This will likely remove your i18n headers because the documentation code replaces the response object with a new response object: 
```
...
setAll(cookiesToSet) {
  cookiesToSet.forEach(({ name, value, options }) => request.cookies.set(name, value))
  supabaseResponse = NextResponse.next({
    request,
  })
  cookiesToSet.forEach(({ name, value, options }) =>
    supabaseResponse.cookies.set(name, value, options)
  )
}, 
...

**Reply by jesper-bylund** [reply]
@brechtpallemans Very interesting, thank you for sharing this Brecht. Still does not explain the 404 though? In my case the i18n is either in the URL or headers, and the non i18n URL is still valid. So even without the i18n header it should return a page, just in English.

**Comment by tretail** [comment]
Experiencing the same issue:
- Self hosted, not Vercel-related
- Using next-intl
- Using Supabase

Followed next-intl/supabase setup: https://next-intl.dev/docs/routing/middleware#example-integrating-with-supabase-authentication

Debugging was tricky, this is the way to reproduce:
- Add logs in middlewares to see cookies and headers
- next dev
- https://supabase.com/dashboard/project/***/settings/api#jwt-settings set expiration time to 10 seconds
- log out, login to get a fresh token
- wait 10 sec and refresh = 404
- Every time you wait at least 10 sec to refresh, you'll get a 404

NextJS/Supabase example is making a fresh new response (good practice, does not mutate original response), dropping both headers and cookies from the original response. Cookies are then reset in the response, but not the headers.

```typescript
import { createServerClient } from "@supabase/ssr";
import { type NextRequest, NextResponse } from "next/server";

export async function updateSession(
  request: NextRequest,
  response: NextResponse,
) {

  const i18nHeader = response.headers.get("x-middleware-rewrite");  # Fix part 1: save the header set by next-intl
  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return request.cookies.getAll();
        },
        setAll(cookiesToSet) {
          cookiesToSet.forEach(({ name, value }) =>
            request.cookies.set(name, value),
          );
          response = NextResponse.next({
            request,
          });
          if (i18nHeader) {
            // Set the header again
            response.headers.set("x-middleware-rewrite", i18nHeader);
          }
          cookiesToSet.forEach(({ name, value, options }) =>
            response.cookies.set(name, value, options),
          );
        },
      },
    },
  );

  // Refresh session if expired - required for Server Components
  // https://supabase.com/docs/guides/auth/server-side/nextjs
  await supabase.auth.getUser();

  return response;
}
```
